### PR TITLE
fix Log attempts to output resp.Body #104184

### LINKED ISSUE #104184: Log attempts to output resp.Body
#### What happened:

Log messages like `body of failing http response: &{0x531900 0xc0010d2780 0x775180}` are not very informative.

/sig cloud-provider

#### What you expected to happen:

Either actual response body or none of it if this is a security concern.

/sig security

sig security to confirm it is ok to change this to log an actual response body.

#### How to reproduce it (as minimally and precisely as possible):

two places it is being logged:

- https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/legacy-cloud-providers/gce/gcpcredential/credentialutil.go#L63
- https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/credentialprovider/config.go#L206

#### Environment:
- Kubernetes version (use `kubectl version`):1.20, but doesn't matter
- Cloud provider or hardware configuration:GKE
- OS (e.g: `cat /etc/os-release`):COS

### PR DESCRIPTION
**Pull request by feeltimeQ**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Log messages like body of failing http response: &{0x531900 0xc0010d2780 0x775180} are not very informative.

#### Which issue(s) this PR fixes:

Fixes #104184 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Review by libaoan — APPROVED** [review]

File: pkg/credentialprovider/config.go [hunk 1]
```diff
@@ -202,20 +202,20 @@ func ReadURL(url string, client *http.Client, header *http.Header) (body []byte,
...
 
+	limitedReader := &io.LimitedReader{R: resp.Body, N: maxReadLength}
+	contents, err := ioutil.ReadAll(limitedReader)
+	if err != nil {
+		return nil, err
+	}
+
 	if resp.StatusCode != http.StatusOK {
-		klog.V(2).Infof("body of failing http response: %v", resp.Body)
+		klog.V(2).Infof("body of failing http response: %v", contents)
```
Comment (jyz0309) [inline pkg/credentialprovider/config.go:212]: As the [comment](https://github.com/kubernetes/kubernetes/issues/104184#issuecomment-893935406) said , I think we can move this to structured log~

File: staging/src/k8s.io/legacy-cloud-providers/gce/gcpcredential/credentialutil.go [hunk 2]
```diff
@@ -59,20 +59,20 @@ func ReadURL(url string, client *http.Client, header *http.Header) (body []byte,
...
 
+	limitedReader := &io.LimitedReader{R: resp.Body, N: maxReadLength}
+	contents, err := ioutil.ReadAll(limitedReader)
+	if err != nil {
+		return nil, err
+	}
+
 	if resp.StatusCode != http.StatusOK {
-		klog.V(2).Infof("body of failing http response: %v", resp.Body)
+		klog.V(2).Infof("body of failing http response: %v", contents)
```
Comment (jyz0309) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gcpcredential/credentialutil.go:69]: ditto.

File: pkg/credentialprovider/config.go (hunk 1 above)
Comment (jyz0309) [reply pkg/credentialprovider/config.go:212]: Just like use `klog.V(2).InfoS("...")`

Comment (yuzhiquan) [reply pkg/credentialprovider/config.go:212]:
```suggestion
		klog.V(2).InfoS("Failing http response", "responseBody", contents)
```

File: staging/src/k8s.io/legacy-cloud-providers/gce/gcpcredential/credentialutil.go [hunk 3]
```diff
@@ -59,20 +59,20 @@ func ReadURL(url string, client *http.Client, header *http.Header) (body []byte,
 	}
 	defer resp.Body.Close()
 
+	limitedReader := &io.LimitedReader{R: resp.Body, N: maxReadLength}
```
Comment (yuzhiquan) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gcpcredential/credentialutil.go:62]: `InfoS` can handle bytes display autoly, may be trans this to string is not necessary.

File: pkg/credentialprovider/config.go [hunk 4]
```diff
@@ -202,20 +202,20 @@ func ReadURL(url string, client *http.Client, header *http.Header) (body []byte,
 	}
 	defer resp.Body.Close()
 
+	limitedReader := &io.LimitedReader{R: resp.Body, N: maxReadLength}
```
Comment (jyz0309) [inline pkg/credentialprovider/config.go:205]: InfoS can display []byte, so I think we don't need this change.
Rest LGTM

File: pkg/credentialprovider/config.go [hunk 5]
```diff
@@ -202,20 +202,20 @@ func ReadURL(url string, client *http.Client, header *http.Header) (body []byte,
...
+	contents, err := ioutil.ReadAll(limitedReader)
+	if err != nil {
+		return nil, err
+	}
+
 	if resp.StatusCode != http.StatusOK {
-		klog.V(2).Infof("body of failing http response: %v", resp.Body)
+		klog.V(2).InfoS("Failing http response", "responseBody", contents)
```
Comment (cheftako) [inline pkg/credentialprovider/config.go:212]: Would like to see security view of if its ok to log the response body.

**Comment by cheftako** [comment]
/lgtm
other than slight concern about what might be in the response body.

**Comment by yuzhiquan** [comment]
OK with this.
/lgtm

File: pkg/credentialprovider/config.go [hunk 6]
```diff
@@ -203,7 +203,7 @@ func ReadURL(url string, client *http.Client, header *http.Header) (body []byte,
 	defer resp.Body.Close()
 
 	if resp.StatusCode != http.StatusOK {
-		klog.V(2).Infof("body of failing http response: %v", resp.Body)
+		klog.V(2).InfoS("Failing http response", "responseBody", resp.Body)
```
Comment (liggitt) [inline pkg/credentialprovider/config.go:206]: I wouldn't log the response body for these errors, things like that have resulted in SSRF issues in the past

Comment (pacoxu) [reply pkg/credentialprovider/config.go:206]: /sig security
@SergeyKanzhelev  mentions below
> sig security to confirm it is ok to change this to log an actual response body.

Or we may change it to:
```
klog.V(2).InfoS("Failed to read URL", "statusCode", resp.StatusCode, "URL", url)
```
like https://github.com/kubernetes/kubernetes/commit/0c5d9ed0d21f0ebace1b5bd437077bc6f6e070f2#diff-f9f30fa2d7430c70760d157077317b55246295b19a6563437aa81f4a736fcaebR81-R83

https://github.com/kubernetes/kubernetes/blob/0c5d9ed0d21f0ebace1b5bd437077bc6f6e070f2/pkg/credentialprovider/config.go#L80-L85

Comment (liggitt) [reply pkg/credentialprovider/config.go:206]: +1 to log status code and url, but not response body

**Comment by liggitt** [comment]
/unassign

feel free to reassign to me once this has been reviewed to put it back in my queue

**Comment by pacoxu** [comment]
/lgtm
/retest
/assign liggitt  
after the pr was updated 😄 thanks

**Pull request merged**